### DISCUSSION TITLE
Changing Which-Key / Underlying Keymaps

### DISCUSSION THREAD

**Discussion by karimlevallois** [LazyVim/LazyVim#2588] [Q&A]
Hi everyone, I'm really having issues changing a couple of keybindings.

Basically I want to change `<leader><tab>` to `<leader>t`.

<img width="1680" alt="Screenshot 2024-02-22 at 08 04 52" src="https://github.com/LazyVim/LazyVim/assets/80129273/b238537c-e80c-4cba-866b-2f35abfc2a7a">

...and inside the +tabs I want to change `<leader><Tab>` to `<leader>n`.

<img width="1680" alt="Screenshot 2024-02-22 at 08 05 02" src="https://github.com/LazyVim/LazyVim/assets/80129273/dd2f08ac-98a2-430f-b6d5-7b83d3e57645">

Any help would be greatly appreciated as I have tried many combinations of adding to both my keycaps.lua and which_key.lua.

**Comment by dpetka2001** [comment]
This can't be done easily. 

First option would be to redefine the `opts.defaults` of `which-key` to return the values that you want. But then you would also have to delete all the default LazyVim keymaps for tabs (so that a `which-key` group for `<Tab>` doesn't get created with an empty prefix) and add the keymaps that you want with `<leader>t` equivalent in your `keymaps.lua`.

Second option would be to use a plugin created by another user of LazyVim community that allows users to change the top-level keys of `which-key`. The plugin is called [lazyvim-menu-addon](https://github.com/abeldekat/lazyvim-menu-addon) and the idea was inspired by this [issue](https://github.com/LazyVim/LazyVim/issues/2193).

**Reply by karimlevallois** [reply]
OK, so I actually tried a similar thing myself thinking this may be the case, but obviously didn't have my keymaps.lua and which_key.lua correct. How would I go disabling the current keymaps for tabs? I've seen mixed messages online, some say to add = nil after the keymap and some say to add false. I've tried both, but neither worked for me. May need help on the which-key side to be honest.

Sorry for the questions, I'm moving from LunarVim and trying to convert my configs over and obviously very different.

Also have a question about adding the LSP details (Python, Ruff, Black etc.) to lualine too, but will open up another discussion for that one.

**Reply by dpetka2001** [reply]
You could delete the current keymaps in your own `/lua/config/keymaps.lua` file using `vim.keymap.del`. See the current keymaps [here](https://github.com/LazyVim/LazyVim/blob/91126b9896bebcea9a21bce43be4e613e7607164/lua/lazyvim/config/keymaps.lua#L165-L171).
For `which-key` something like the following should do what you want
```lua
-- ~/.config/nvim/lua/plugins/which-key.lua
return {
  {
    "folke/which-key.nvim",
    opts = function(_, opts)
      opts.defaults["<leader><tab>"] = nil
      opts.defaults["<leader>t"] = { name = "+tabs" }
    end,
  },
}
```

**Reply by karimlevallois** [reply]
```
local del = vim.keymap.del

del("n", "<leader><tab>l", "<cmd>tablast<cr>", { desc = "Last Tab" })
del("n", "<leader><tab>f", "<cmd>tabfirst<cr>", { desc = "First Tab" })
del("n", "<leader><tab><tab>", "<cmd>tabnew<cr>", { desc = "New Tab" })
del("n", "<leader><tab>]", "<cmd>tabnext<cr>", { desc = "Next Tab" })
del("n", "<leader><tab>d", "<cmd>tabclose<cr>", { desc = "Close Tab" })
del("n", "<leader><tab>[", "<cmd>tabprevious<cr>", { desc = "Previous Tab" })

map("n", "<leader>tl", "<cmd>tablast<cr>", { desc = "Last Tab" })
map("n", "<leader>tf", "<cmd>tabfirst<cr>", { desc = "First Tab" })
map("n", "<leader>tn", "<cmd>tabnew<cr>", { desc = "New Tab" })
map("n", "<leader>t]", "<cmd>tabnext<cr>", { desc = "Next Tab" })
map("n", "<leader>td", "<cmd>tabclose<cr>", { desc = "Close Tab" })
map("n", "<leader>t[", "<cmd>tabprevious<cr>", { desc = "Previous Tab" })
```

in keycaps.lua?

Getting an error saying expecting table, got string.

**Reply by dpetka2001** [reply]
It should be just `del("n", "<leader><tab>l")`. See `:h vim.keymap.del()` for more info.
